File: plugin/testData/shared/test.shared.kt
package com.lightningkite.kwift.async

import com.lightningkite.kwift.escaping

typealias DRF<T> = DelayedResultFunction<T>
inline class DelayedResultFunction<T>(val value: @escaping() (@escaping() (T) -> Unit) -> Unit) {
    fun invoke(callback: @escaping() (T) -> Unit) {
        value(callback)
    }
}

fun <T> immediate(action: @escaping() ()->T): DelayedResultFunction<T> {
    return DelayedResultFunction { onResult -> onResult(action()) }
}
fun <T> drfStart(value: T): DelayedResultFunction<T> = DelayedResultFunction { onResult -> onResult(value) }

fun <T, B> DelayedResultFunction<T>.then(next: (T) -> DelayedResultFunction<B>): DelayedResultFunction<B> {
    val first = this
    return DelayedResultFunction { onResult ->
        first.invoke { input ->
            next(input).invoke(onResult)
        }
    }
}
fun <T, B> DelayedResultFunction<T>.thenImmediate(next: (T) -> B): DelayedResultFunction<B> {
    val first = this
    return DelayedResultFunction<B> { onResult ->
        first.invoke { input ->
            onResult(next(input))
        }
    }
}

fun test() {
    drfStart(2)
        .then { a -> immediate { 4 + a } }
        .thenImmediate { a -> 4 + a }
        .then { a -> DelayedResultFunction<Double> { onResult -> onResult(a.toDouble() + 4.0) } }
        .invoke {
            println(it)
        }
}

File: plugin/testDataOutput/shared/test.shared.swift
//Package: com.lightningkite.khrysalis.async
//Converted using Khrysalis2

import Foundation
import RxSwift
import RxRelay



public typealias DRF<T> = DelayedResultFunction<T>
 

public class DelayedResultFunction<T> {
    
    public var value:  (@escaping (T) -> Void) -> Void
    
    
    public func invoke(callback: @escaping (T) -> Void) -> Void {
        value(callback)
    }
    
    public init(value: @escaping (@escaping (T) -> Void) -> Void) {
        self.value = value
    }
    convenience public init(_ value: @escaping (@escaping (T) -> Void) -> Void) {
        self.init(value: value)
    }
}
 
 

public func immediate<T>(action: @escaping () -> T) -> DelayedResultFunction<T> {
    return DelayedResultFunction{ (onResult) in 
        onResult(action())
    }
}
 

public func drfStart<T>(value: T) -> DelayedResultFunction<T> {
    return DelayedResultFunction{ (onResult) in 
        onResult(value)
    }
}
public func drfStart<T>(_ value: T) -> DelayedResultFunction<T> {
    return drfStart(value: value)
}
 
 

extension DelayedResultFunction {
    public func then<B>(next: (T) -> DelayedResultFunction<B>) -> DelayedResultFunction<B> {
        var first = self
        return DelayedResultFunction{ (onResult) in 
            first.invoke{ (input) in 
                next(input).invoke(onResult)
            }
        }
    }
}
 

extension DelayedResultFunction {
    public func thenImmediate<B>(next: (T) -> B) -> DelayedResultFunction<B> {
        var first = self
        return DelayedResultFunction < B > { (onResult) in 
            first.invoke{ (input) in 
                onResult(next(input))
            }
        }
    }
}
 
 

public func test() -> Void {
    drfStart(2).then{ (a) in 
        immediate{ () in 
            4 + a
        }
    }.thenImmediate{ (a) in 
        4 + a
    }.then{ (a) in 
        DelayedResultFunction{ (onResult) in 
            onResult(a.toDouble() + 4.0)
        }
    }.invoke{ () in 
        print(it)
    }
}
 
